command/aarelayer/service/aa_types.go

```diff
@@ -42,54 +41,118 @@ var (
 		"tuple(bytes32 typeHash, bytes32 name, bytes32 version, uint256 chainId, address verifyingContract)",
 	)
 
+	// invokerMethodAbiType is function invoke(
+	//	Signature calldata signature, Transaction calldata transaction) external payable nonReentrant
+	invokerMethodAbiType = abi.MustNewMethod(
+		`function invoke(
+			tuple(uint256 r,uint256 s,bool v) signature,
+			tuple(address from,uint256 nonce,tuple(address to,uint256 value,uint256 gasLimit,bytes data)[] payload) transaction
+		) public payable`,
+	)
+
 	//  aaInvokerNoncesAbiType is mapping(address => uint256) public nonces;
 	aaInvokerNoncesAbiType = abi.MustNewMethod("function nonces(address) returns (uint256)")
 )
 
+type MagicHashFn func(chainID int64, address types.Address, commitHash []byte) []byte
+
 // AATransaction represents an AA transaction
 type AATransaction struct {
 	Signature   aaSignature `json:"signature"`
 	Transaction Transaction `json:"transaction"`
 }
 
-func (t *AATransaction) GetAddressFromSignature(address types.Address, chainID int64) types.Address {
+func (t *AATransaction) RecoverSender(
+	address types.Address, chainID int64, magicHashFn MagicHashFn) types.Address {
 	domainSeparator, err := GetDomainSeperatorHash(address, chainID)
 	if err != nil {
 		return types.ZeroAddress
 	}
 
-	hash, err := t.Transaction.ComputeEip712Hash(domainSeparator)
+	commitHash, err := t.Transaction.ComputeEip712Hash(domainSeparator)
```

Comment:
Add comments to the `ComputeEip712Hash` function. Also probably would be good to extract to some variable this line:
`headerBytes := [2]byte{0x19, 0x1}`, or at least provide a link to the EIP712, which explains it in details.